### PR TITLE
After calling nextHandler(), backspace completion should verify that …

### LINKED ISSUE #3789: ModelComputer throws InvalidOperationException when editing a diff document
When editing a diff document, if I press Delete on the line just above the beginning of a red (removed) region, I get the following error.

image

+		$exception	{"Nullable object must have a value."}	System.InvalidOperationException

	mscorlib.dll!System.ThrowHelper.ThrowInvalidOperationException(System.ExceptionResource resource)	Unknown
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.ModelComputer.ModelComputer(Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session session, Microsoft.CodeAnalysis.Completion.ICompletionService completionService, Microsoft.CodeAnalysis.Completion.CompletionTriggerInfo triggerInfo, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Completion.Providers.ICompletionProvider> completionProviders, bool isDebugger) Line 80	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Session.ComputeModel(Microsoft.CodeAnalysis.Completion.ICompletionService completionService, Microsoft.CodeAnalysis.Completion.CompletionTriggerInfo triggerInfo, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Completion.Providers.ICompletionProvider> completionProviders, bool isDebugger) Line 40	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.StartNewModelComputation(Microsoft.CodeAnalysis.Completion.ICompletionService completionService, Microsoft.CodeAnalysis.Completion.CompletionTriggerInfo triggerInfo, bool filterItems, bool dismissIfEmptyAllowed) Line 174	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.ExecuteBackspaceOrDelete(Microsoft.VisualStudio.Text.Editor.ITextView textView, System.Action nextHandler, bool isDelete) Line 107	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.Controller.Microsoft.CodeAnalysis.Editor.ICommandHandler<Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs>.ExecuteCommand(Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action nextHandler) Line 24	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.CommandHandlers.AbstractCompletionCommandHandler.ExecuteCommandWorker<Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs>(Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action nextHandler) Line 87	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.CommandHandlers.AbstractCompletionCommandHandler.ExecuteCommand(Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action nextHandler) Line 312	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Commands.CommandHandlerService.ExecuteHandlers.AnonymousMethod__1() Line 93	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.HandlePossibleTypingCommand(Microsoft.CodeAnalysis.Editor.CommandArgs args, System.Action nextHandler, System.Action<Microsoft.VisualStudio.Text.SnapshotSpan> actionIfInsideActiveSpan) Line 82	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.InlineRename.RenameCommandHandler.ExecuteCommand(Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action nextHandler) Line 44	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Commands.CommandHandlerService.ExecuteHandlers<Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs>(System.Collections.Generic.IList<Microsoft.CodeAnalysis.Editor.ICommandHandler<Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs>> commandHandlers, Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action lastHandler) Line 107	C#
	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Implementation.Commands.CommandHandlerService.Execute<Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs>(Microsoft.VisualStudio.Utilities.IContentType contentType, Microsoft.CodeAnalysis.Editor.Commands.DeleteKeyCommandArgs args, System.Action lastHandler) Line 71	C#
	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.ExecuteDelete(Microsoft.VisualStudio.Text.ITextBuffer subjectBuffer, Microsoft.VisualStudio.Utilities.IContentType contentType, System.Action executeNextCommandTarget) Line 714	C#
	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.ExecuteVisualStudio97(ref System.Guid pguidCmdGroup, uint commandId, uint executeInformation, System.IntPtr pvaIn, System.IntPtr pvaOut, Microsoft.VisualStudio.Text.ITextBuffer subjectBuffer, Microsoft.VisualStudio.Utilities.IContentType contentType) Line 184	C#
	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.AbstractOleCommandTarget.Exec(ref System.Guid pguidCmdGroup, uint commandId, uint executeInformation, System.IntPtr pvaIn, System.IntPtr pvaOut) Line 53	C#
	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.CommandChainNode.InnerExec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.CommandChainNode.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Diff.dll!Microsoft.VisualStudio.Differencing.Package.BlockingCommandFilter.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.SimpleTextViewWindow.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Editor.Implementation.dll!Microsoft.VisualStudio.Editor.Implementation.CompoundTextViewWindow.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Diff.dll!Microsoft.VisualStudio.Differencing.Package.DiffViewerCodeWindow.Microsoft.VisualStudio.OLE.Interop.IOleCommandTarget.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Platform.WindowManagement.dll!Microsoft.VisualStudio.Platform.WindowManagement.DocumentObjectSite.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
	Microsoft.VisualStudio.Platform.WindowManagement.dll!Microsoft.VisualStudio.Platform.WindowManagement.WindowFrame.Exec(ref System.Guid pguidCmdGroup, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown

D14Rel 23024.00
git commit baseline: VS 9b99a170666

### PR DESCRIPTION
**Pull request by rchande**
…the caret can still be mapped into our subject buffer before starting a session.

Fixes #3789.

cc @jasonmalinowski @dpoeschl @balajikris @brettfo @DustinCampbell @CyrusNajmabadi @davkean @jmarolf @Pilchie

### CODE REVIEW

**Comment by dpoeschl** [comment]
Unit test?

**Comment by dpoeschl** [comment]
Can other commands cause this?

**Comment by Pilchie** [comment]
+1 to both of @dpoeschl's questions :smile:

**Comment by jasonmalinowski** [comment]
:+1:, modulo @dpoeschl's questions. :-)

**Comment by rchande** [comment]
I moved the check up into StartModelComputation itself, and added a test.

**Comment by rchande** [comment]
And yes, @dpoeschl, I verified that the test passes without the caret mapping check :smile:

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 1]
```diff
@@ -1507,5 +1514,44 @@ $$]]></Document>, extraExportedTypes:={GetType(CSharpEditorFormattingService)}.T
                 state.AssertMatchesTextStartingAtLine(6, "doo;")
             End Using
         End Sub
+
+        <WorkItem(588, "https://github.com/dotnet/roslyn/issues/588")>
```
Comment (dpoeschl) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: 3789, not 588

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 2]
```diff
@@ -1507,5 +1514,44 @@ $$]]></Document>, extraExportedTypes:={GetType(CSharpEditorFormattingService)}.T
...
+class C
+{
+    void foo(int x)
+    {$$
+        /********/
+        int doodle;
+        }
+}]]></Document>, extraExportedTypes:={GetType(CSharpEditorFormattingService)}.ToList())
+
+                Dim differenceBufferService = state.GetExportedValue(Of IDifferenceBufferFactoryService)
```
Comment (dpoeschl) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: Unused?

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 3]
```diff
@@ -1507,5 +1514,44 @@ $$]]></Document>, extraExportedTypes:={GetType(CSharpEditorFormattingService)}.T
...
+                Dim textBufferFactoryService = state.GetExportedValue(Of ITextBufferFactoryService)()
```
Comment (dpoeschl) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: Also unused.

File: src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb [hunk 4]
```diff
@@ -1507,5 +1514,44 @@ $$]]></Document>, extraExportedTypes:={GetType(CSharpEditorFormattingService)}.T
...
+                Dim contentTypeService = state.GetExportedValue(Of IContentTypeRegistryService)()
+                Dim contentType = contentTypeService.GetContentType(ContentTypeNames.CSharpContentType)
+                Dim textViewFactory = state.GetExportedValue(Of ITextEditorFactoryService)()
+                Dim editorOperationsFactory = state.GetExportedValue(Of IEditorOperationsFactoryService)()
+
+                Dim otherBuffer = textBufferFactoryService.CreateTextBuffer("text", contentType)
+                Dim otherExposedSpan = otherBuffer.CurrentSnapshot.CreateTrackingSpan(0, 4, SpanTrackingMode.EdgeExclusive, TrackingFidelityMode.Forward)
```
Comment (dpoeschl) [inline src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb]: A few comments would go a long way toward explaining what you're trying to pull off here.

**Comment by rchande** [comment]
Looks like something went wrong with the Linux build.

**Pull request merged**